rust_src/remacs-sys/lib.rs

```diff
@@ -665,6 +665,30 @@ pub struct Lisp_String {
     pub data: *mut libc::c_char,
 }
 
+#[repr(C)]
+pub union SymbolUnion {
+    pub value: Lisp_Object,
+    pub alias: *mut Lisp_Symbol,
+    pub blv: *mut libc::c_void, // @TODO implement Lisp_Buffer_Local_Value
+    pub fwd: *mut libc::c_void, // @TODO implement Lisp_Fwd
+}
+
+/// This struct has 4 bytes of padding, representing the bitfield that 
+/// lives at the top of a Lisp_Symbol. The first 10 bits of this field are
+/// used
+
+// @TODO check the value of name post and pre transmutation, it seems that name is surviving but
+// may not be the correct value
+#[repr(C)]
+pub struct Lisp_Symbol {
```

Comment:
So this is my understanding of the situation re: the mentioned variables. If anyone spots any incorrect information please do not hesitate to correct me:

Lisp_Symbol has the following definition in lisp.h (I've stripped out the comments for clarity):
```
struct Lisp_Symbol
{
  bool_bf gcmarkbit : 1;
  ENUM_BF (symbol_redirect) redirect : 3;
  ENUM_BF (symbol_trapped_write) trapped_write : 2;
  unsigned interned : 2;
  bool_bf declared_special : 1;
  bool_bf pinned : 1;
  Lisp_Object name;
  union {
    Lisp_Object value;
    struct Lisp_Symbol *alias;
    struct Lisp_Buffer_Local_Value *blv;
    union Lisp_Fwd *fwd;
  } val;
  Lisp_Object function;
  Lisp_Object plist;
  struct Lisp_Symbol *next;
```
This struct is using the C notation for bit fields. According to my research, Rust does not (https://github.com/rust-lang/rfcs/issues/314, https://users.rust-lang.org/t/c-structs-with-bit-fields-and-ffi/1429)  support setting up an equivalent for C bitfieldsin it's `#[repr(C)]` directive.

On my system (64-bit Ubuntu 16.04), no matter what the typedef of bf_bool or ENUM_BF, the bit field section of the struct takes 4 bytes. Due to alignment, the struct is padded, and offsetof(struct Lisp_Symbol, name) reports 8. My initial solution to representing this struct in Rust was to represent the bit field block as a u32, taking up the 4 bytes I mentioned earlier. I have not fully convinced myself that this is 100% the correct and portable thing to do. 

Even if we can safely represent this block with a u32 on every system we support, it seems that there are 'gotcha's with accessing these bit fields via the bit wise operators (due to endianness, and compiler differences w.r.t bit field implementation.) 

Overall I am not sure the best way to handle the interop for C structs that use bit fields that we need to access in Rust. It seems that one option is to simply pad the Rust struct as best we can, and if we need to access these fields, we will need to maintain C bindings that access them for us.